### DISCUSSION TITLE
Attribute doubts/proposals

### DISCUSSION THREAD

**Discussion by jugaadi** [cuelang/cue#1009] [Ideas]
1. Any reason for not allowing attributes at the top of fields/struct/package identifiers? For multi-line string, putting an attribute at the end of the field affects the readability.
2. Can I use cue inside an attribute as part of its content? This will enable reuse of the content.
3. Should we reserve attribute names based on file types? So `@py` will always represent python
(edited)

**Comment by myitcv** [comment]
@jugaadi - thank you for re-raising this on GitHub discussions. 

> Any reason for not allowing attributes at the top of fields/struct/package identifiers? For multi-line string, putting an attribute at the end of the field affects the readability.

You can use embedded scalars to make this a bit more readable:

```
f: {
	@hello(This is
		a story about a very
		big attribute message
		)
	4
}
```

If there are other situations that feel awkward, please can you share them?



> Can I use cue inside an attribute as part of its content? This will enable reuse of the content.

Attributes [do not influence the evaluation of CUE](https://tip.cuelang.org/docs/references/spec/#attributes). They exist to allow associating meta information with values.

> Should we reserve attribute names based on file types? So `@py` will always represent python

This is a really good question that I think we should discuss in the form of an issue if you don't mind creating an issue. Certain attributes are already "reserved" by virtue of their use in the CUE distribution (e.g. `@protobuf`). Should the CUE distribution reserve others: possibly? Will we be able to enforce this? Probably not. But still worth discussing.

**Reply by jugaadi** [reply]
Thanks @myitcv  for your response. 

Consider this cue:
```
_args: {
    id: 23,
    bar: "baz"
}

f: {
	@hello(_args)
	
	j: """
	 Lorem ipsum dolor sit amet, consectetur adipiscing elit,
	 sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 
	 Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi 
	 ut aliquip ex ea  commodo consequat. Duis aute irure dolor in 
	 reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla 
	 pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa 
	 qui officia deserunt mollit anim id est laborum. """  @go(Field, j=_args)
}
```
1. In the above example, the `@go` attribute is a bit unreadable. So if we put on top of the fields like the example given below, it would be helpful.
```
@hello(_args)
f: {
	
	@go(Field, j=_args)
	j: """
	 Lorem ipsum dolor sit amet, consectetur adipiscing elit,
	 sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 
	 Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi 
	 ut aliquip ex ea  commodo consequat. Duis aute irure dolor in 
	 reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla 
	 pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa 
	 qui officia deserunt mollit anim id est laborum. """
}
```
2. Similarly, I used `_args` as part of content in  `@hello` as well as `@go` attributes. This helps use(and reuse) of structured data or interpolated strings.

3. Sure.. Will create an issue.

**Reply by myitcv** [reply]
Regarding 1. I'll let @mpvl confirm for sure, but I doubt we will move to a model of allowing attributes to be declared in that way. Not least because it would create an ambiguity: does it apply to the field that follows or the struct within which it is declared?

I would rewrite your example as follows:

```
_args: {
	id:  23
	bar: "baz"
}

f: {
	@hello(_args)

	j: {
		@go(Field, j=_args)
		"""
			Lorem ipsum dolor sit amet, consectetur adipiscing elit,
			sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
			Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi
			ut aliquip ex ea  commodo consequat. Duis aute irure dolor in
			reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
			pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa
			qui officia deserunt mollit anim id est laborum.
			"""
	}
}
```

Which is slightly better.

I have to admit I'm a bit lost regarding what you mean for point 2. Attributes are largely entirely free-form (the only constraint being balanced brackets). Certain attributes (`@go`, `@protobuf`... ) have meaning to core CUE, but beyond that you can do what you like. Can you explain a bit more what you mean here?

Regarding point 3: thanks.

**Reply by jugaadi** [reply]
1. It should be applied to the field/struct/package that immediately follows it or the attribute below it.
2. 
Output
```xml
<a x="Hello World" y=5>Foobar</a>
```

Input1: 
```
_m: 4
_n: 1
_o: "World"

a: Foobar @xml(x="Hello \(_o)" y=\(m+n))
```

or

Input2: 
```
_a: {
    x: "Hello World"
    y: 5
}
a: Foobar @xml(_a)
```

Here I would like to use cue expressions inside the attribute. Or did I misunderstand the usage?

**Reply by myitcv** [reply]
> 1. It should be applied to the field/struct/package that immediately follows it or the attribute below it.

That's problematic for the reason I explained above.

> Here I would like to use cue expressions inside the attribute. Or did I misunderstand the usage?

From [the spec](https://tip.cuelang.org/docs/references/spec/#attributes):

> Attributes allow associating meta information with values. Their primary purpose is to define mappings between CUE and other representations. Attributes do not influence the evaluation of CUE.

Hence, you cannot use a CUE value in an attribute. Attributes exist outside of CUE evaluation.

**Reply by jugaadi** [reply]
Thanks @myitcv for your response. Should I create an issue(Feature Request) for the above two?

**Reply by myitcv** [reply]
By all means create feature requests, but as I've outlined I would be surprised if we ever move to allow attributes to be defined before an identifier; similarly, it's feature of attributes that they exist outside of CUE evaluation.

**Reply by jugaadi** [reply]
For (3), I have created an issue here - https://github.com/cuelang/cue/issues/1019

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/1009.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
